extension/background.js

```diff
@@ -0,0 +1,199 @@
+      [TELEMETRY_METHOD_ETLD_CHANGE]: {
+        methods: [TELEMETRY_METHOD_ETLD_CHANGE],
+        objects: [TELEMETRY_OBJECT_WEBREQUEST, TELEMETRY_OBJECT_OTHER],
+        extra_keys: ["addonId", "addonVersion", "from", "to"],
+        record_on_release: true,
+      },
+    });
+
+    this.onRedirectedListener = this.onRedirectedListener.bind(this);
+  }
+
+  async getMatchPatterns() {
+    try {
+      this.matchPatterns =
+        await browser.addonsSearchExperiment.getMatchPatterns();
+    } catch (err) {
+      console.error(`failed to retrieve the list of URL patterns: ${err}`);
+      this.matchPatterns = {};
+    }
+
+    return this.matchPatterns;
+  }
+
+  // When the search service changes the set of engines that are enabled, we
+  // update our pattern matching in the webrequest listeners (go to the bottom
+  // of this file for the search service events we listen to).
+  async monitor() {
+    // If there is already a listener, remove it so that we can re-add one
+    // after. This is because we're using the same listener with different URL
+    // patterns (when the list of search engines changes).
+    if (
+      browser.addonsSearchExperiment.onRedirected.hasListener(
+        this.onRedirectedListener
+      )
+    ) {
+      this.debug("removing onRedirected listener");
+      browser.addonsSearchExperiment.onRedirected.removeListener(
+        this.onRedirectedListener
+      );
+    }
+    // If there is already a listener, remove it so that we can re-add one
+    // after. This is because we're using the same listener with different URL
+    // patterns (when the list of search engines changes).
+    if (browser.webRequest.onBeforeRequest.hasListener(this.noOpListener)) {
+      this.debug("removing onBeforeRequest listener");
+      browser.webRequest.onBeforeRequest.removeListener(this.noOpListener);
+    }
+
+    // Retrieve the list of URL patterns to monitor with our listener.
+    //
+    // Note: search suggestions are system principal requests, so webRequest
+    // cannot intercept them.
+    const matchPatterns = await this.getMatchPatterns();
+    const patterns = Object.keys(matchPatterns);
+
+    if (patterns.length === 0) {
+      this.debug(
+        "not registering any listener because there is no URL to monitor"
+      );
+      return;
+    }
+
+    this.debug("registering onBeforeRequest listener");
+    browser.webRequest.onBeforeRequest.addListener(
+      this.noOpListener,
+      { types: ["main_frame"], urls: patterns },
+      ["blocking"]
+    );
+
+    this.debug("registering onRedirected listener");
+    browser.addonsSearchExperiment.onRedirected.addListener(
+      this.onRedirectedListener,
+      { urls: patterns }
+    );
+  }
+
+  // This listener is required to force the registration of traceable channels.
+  noOpListener() {
+    // Do nothing.
+  }
+
+  async onRedirectedListener({ addonId, firstUrl, lastUrl }) {
+    if (!firstUrl || !lastUrl) {
```

Comment:
Any reason for having this check here instead of api.js?

Would it make sense to move this check (and the generation of the public suffix) to the parent? The only thing that can potentially depend on the full URL is the generation of `addonIds`, but that can/should probably also be in the api.js script as mentioned elsewhere (search for "matched URL" in my review).

Reply:
I moved this check to the api code to avoid firing an event. The rest is kept in the background script, which describes the reporting logic.